### PR TITLE
Save state

### PR DESCRIPTION
**Pull request by vogievetsky**

### CODE REVIEW

File: src/client/visualizations/line-chart/line-chart.tsx [hunk 1]
```diff
@@ -553,7 +554,23 @@ export class LineChart extends BaseVisualization<LineChartState> {
         newState.continuousDimension = continuousDimension;
 
         var axisRange = essence.getEffectiveFilter(LineChart.id).getExtent(continuousDimension.expression) as PlywoodRange;
-        axisRange = axisRange ? axisRange : this.getXAxisRange(essence, continuousDimension, dataset);
+        if (axisRange) {
+          // Special treatment for realtime data, i.e. time data where the maxTime is within Duration of the filter end
```
Comment (lorem--ipsum) [inline src/client/visualizations/line-chart/line-chart.tsx]: Is it possible to put this into a separate function ? The function's name would act as a comment :)

File: src/common/models/measure/measure.ts [hunk 2]
```diff
@@ -224,22 +232,34 @@ export class Measure implements Instance<MeasureValue, MeasureJS> {
     return this.formatFn(datum[this.name] as number);
   }
 
-  public change(propertyName: string, newValue: any): Measure {
+  public change(propertyName: string, newValue: any): this {
```
Comment (lorem--ipsum) [inline src/common/models/measure/measure.ts]: Return type as `this` ? What is this sorcery ?

File: src/server/config.ts [hunk 3]
```diff
@@ -18,7 +18,7 @@ import * as path from 'path';
 import * as nopt from 'nopt';
 import { arraySum } from '../common/utils/general/general';
 import { Cluster, DataCube, SupportedType, AppSettings } from '../common/models/index';
-import { clusterToYAML, dataCubeToYAML } from '../common/utils/yaml-helper/yaml-helper';
+import { appSettingsToYAML } from '../common/utils/yaml-helper/yaml-helper';
```
Comment (lorem--ipsum) [inline src/server/config.ts]: 👍

File: src/server/config.ts [hunk 4]
```diff
@@ -253,11 +260,11 @@ const CLUSTER_TYPES: SupportedType[] = ['druid', 'postgres', 'mysql'];
 
 var settingsLocation: SettingsLocation = null;
 if (serverSettingsFilePath) {
-  settingsLocation = {
-    location: 'local',
-    readOnly: false, // ToDo: this should be true
-    uri: serverSettingsFilePath
-  };
+  if (SERVER_SETTINGS.settingsUri) {
```
Comment (lorem--ipsum) [inline src/server/config.ts]: I'm thinking maybe a `writable` flag on the AppSettings could be useful ? To put the UI in `read-only` mode or something ? Not necessarily for now though

Comment (vogievetsky) [inline src/server/config.ts]: Just added on to SETTINGS_MANAGER we should talk about read only UI and if we want that

File: src/common/models/measure/measure.ts (hunk 2 above)
Comment (vogievetsky) [inline src/common/models/measure/measure.ts]: yeah it is meant to indicate that the function return something if it's own type. This is useful when when function can be inherited but still return the type of the object being called.

Hold on I will add more info in a sec.

Comment (vogievetsky) [inline src/common/models/measure/measure.ts]: Ok so in: https://www.typescriptlang.org/docs/handbook/advanced-types.html read the section "Polymorphic this types".

Comment (vogievetsky) [inline src/common/models/measure/measure.ts]: Please see hear for real world usage: https://github.com/implydata/immutable-class/pull/1/files#diff-ae89c91267724b7637f7bb0f71f7ffdcR114

I was just trying it out here as a test, eventually this `change` method in Measure will be deleted anyways

Comment (vogievetsky) [inline src/common/models/measure/measure.ts]: Just to be clear this here is a really confusing place to use it, I was just trying it out. see the immutable class PR for a 'real' use case

**Comment by evasomething** [comment]
💧

**Pull request merged**